zeusd/src/devices/cpu/linux.rs

```diff
@@ -0,0 +1,219 @@
+
+        match fs::read_dir(&base_path) {
+            Ok(entries) => {
+                for entry in entries {
+                    if let Ok(entry) = entry {
+                        let path = entry.path();
+                        if path.is_dir() {
+                            if let Some(dir_name_str) = path.file_name() {
+                                let dir_name = dir_name_str.to_string_lossy();
+                                if dir_name.contains("intel-rapl") {
+                                    index_count += 1;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Err(_) => {
+                tracing::error!("RAPL not available");
+            }
+        };
+        Ok(index_count)
+    }
+
+    fn get_available_fields(index: u32) -> Result<(PackageInfo, Option<PackageInfo>), ZeusdError> {
+        let base_path = PathBuf::from(format!("{}/intel-rapl:{}", RAPL_DIR, index));
+        let cpu_info = PackageInfo::new(&base_path, index)?;
+
+        let mut dram_info = None;
+        match fs::read_dir(&base_path) {
+            Ok(entries) => {
+                for entry in entries {
+                    if let Ok(entry) = entry {
+                        let path = entry.path();
+                        if path.is_dir() {
+                            if let Some(dir_name_str) = path.file_name() {
+                                let dir_name = dir_name_str.to_string_lossy();
+                                if dir_name.contains("intel-rapl") {
+                                    let subpackage_path = base_path.join(&*dir_name);
+                                    let subpackage_info =
+                                        PackageInfo::new(&subpackage_path, index)?;
+                                    if subpackage_info.name == "dram" {
+                                        dram_info = Some(subpackage_info);
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Err(_) => {
+                return Err(ZeusdError::CpuInitializationError(
+                    index.try_into().unwrap(),
+                ))
+            }
+        };
+
+        Ok((cpu_info, dram_info))
+    }
+
+    fn get_cpu_energy(&self) -> anyhow::Result<u64, ZeusdError> {
+        let handle = self
+            .cpu_monitoring_task
+            .get_or_init(|| tokio::spawn(monitor_rapl(Arc::clone(&self.cpu))));
+        if handle.is_finished() {
+            return Err(ZeusdError::CpuManagementTaskTerminatedError(
+                self.cpu.index.try_into().unwrap(),
+            ));
+        }
+
+        let measurement = read_u64(&self.cpu.energy_uj_path)?;
+        let num_wraparounds = self.cpu.num_wraparounds.read().unwrap();
```

Comment:
Timing bug?

Order of events:
1. This method executes line 142. `measurement` is a pretty large number, close to max energy uj.
2. A wraparound happens and it is detected by the background monitor. `num_wraparounds` is incremented by 1.
3. This method executes line 143 and reads the incremented `num_wraparounds`.
4. The resulting energy number returned by this method is surprisingly large.

Reply:
Addressing this by getting the wraparound counts before and after the measurement. If the wraparound count is different after the measurement, retake the measurement.